Adds MMProperty class - a view onto a core property with a nicer API

tlambert03 opened the pull request:
this is a convenience class that lets you avoid doing a lot of searching for the property related API on the core object.

```python
In [1]: from pymmcore_plus import MMProperty

In [2]: prop = MMProperty('Objective', 'Label')

In [3]: prop.isValid()  # doesn't need to be loaded in core
Out[3]: False

In [4]: prop.core.loadSystemConfiguration()

In [5]: prop.isValid()  # now it's loaded in core
Out[5]: True

In [6]: prop.value
Out[6]: 'Nikon 10X S Fluor'

In [7]: prop.value = 'Objective-2'  # property.setter

In [8]: prop.value
Out[8]: 'Objective-2'

In [9]: prop.value == prop.core.getProperty(prop.device, prop.name)  # stays in sync
Out[9]: True

In [10]: prop.isReadOnly()
Out[10]: False

In [11]: prop.hasLimits()
Out[11]: False

In [12]: prop.type()
Out[12]: <PropertyType.String: 1>

In [13]: prop  # nice repr
Out[13]: <Property Label on device Objective: value='Objective-2'>
```

ianhi commented on pymmcore_plus/core/_property.py:
```diff
@@ -0,0 +1,145 @@
+
+    >>> prop = MMProperty('Objective', 'Label')
+    >>> prop.dict()
+    >>> prop.core.loadSystemConfiguration()
```
I'm not sure i love this access pattern. I can see having it be allowed on this object but maybe we should try to always use `instance()` in examples?

tlambert03 replied:
can remove.  Ultimately, this class is just a combination of a core instance, and two strings.  I think it makes sense to be able to access and  set the core instance, but this particular example "sneakily" access the global singleton.  Is it just that?  or is it the fact that you can get `prop.core` at all?

ianhi replied:
It's just the sneaky access that was worrying me. The functionality of accessing it makes sense

tlambert03 replied:
Then I think we should make core a mandatory parameter, and remove `self._mmc = mmcore or CMMCorePlus.instance()` from line 46

ianhi reviewed (changes requested):
This looks incredible!

My only concern is how to treat the `core` object on these devices. Basically I don't think that `core` should be a settable property, but if it is then `setCore` needs a docstring clarifying what happens (both to the object and the new core) when the core is swapped.

A nice addition to consider (here or follow up) is returning these objects for the `getDeviceProperties` which currently returns a dict with only the current values. I htink that could really increase usability of `getDeviceProperties` but it would involve creating lots of objects.

tlambert03 commented on pymmcore_plus/core/_property.py:
```diff
@@ -0,0 +1,145 @@
+    def isLoaded(self) -> bool:
+        return self._mmc is not None and self.device in self._mmc.getLoadedDevices()
+
+    @property
+    def core(self) -> CMMCorePlus:
+        return self._mmc
+
+    @core.setter
+    def core(self, mmcore: CMMCorePlus) -> None:
+        self._mmc = mmcore
+
+    # functional alternate to property setter
+    def setCore(self, mmcore: CMMCorePlus) -> None:
+        self._mmc = mmcore
```
> Most importantly, if I get a prop from one core, switch it's value, then swap it to another does that new core get sent a setProperty

nope, this important thing to recognize here is that this is a completely stateless class (outside of core, device label and property name).  It simply wraps the api of `core.doAThin(device, property)` ... so if you switch core, you switch what you would get from `core.getAthing(device, property)`.

tlambert03 replied:
(that's really what I meant by "view" here... Perhaps calling this a `PropertyView` would be good?)

ianhi replied:
>  nope, this important thing to recognize here is that this is a completely stateless class 

Ah yes - i wasn't considering that! So i guess my worry is more so that it may be easy to assume that these are stateful object (like i did) and then be surprised when they are not.

Also I hope that people don't have multiple cores lying around, that seems like it would only lead to bad times.

tlambert03 replied:
yeah, fair enough :)  OK, going to make core a mandatory parameter in `__init__` and unsetable

tlambert03 replied:
(if we ever later decide that it would be nice to instantiate one of these objects before you even know about a core, we can change later)

ianhi replied:
Don't forget the middle ground of `None` and defaulting to using `instance` potentially avoids some typing in the hopefully common case of only one core. But I also see how it opens up a box of worms

tlambert03 replied:
I think I'll add a method to CMMCorePlus ... so it would be CMMCorePlus.instance().propertyObject('dev', 'label')

ianhi replied:
> CMMCorePlus.instance().propertyObject('dev', 'label')

What do you think about `getProperty` returning one of these objects?

ianhi replied:
Ahh actually I see how that breaks the symmetry of the `get` and `set` property functions, which would be bad.

tlambert03 replied:
I considered it, but I think that's too much of a divergence from the core API.  There are other cases, like `getConfigData` or other methods that return a `Configuration` where I added the `native=False` parameter, in which case we return an improved object... but those objects simply _add_ to the pymmcore API.   Here, we'd be fully changing the API from something that returns a value, to something that returns something that has a `.value` attribute ... so i think it's too confusing

tlambert03 replied:
not married to `propertyObject` as the method name though!

ianhi replied:
Definitely agree that it would become too confusing. 

I think this is pretty good. Only other ones I might consider are `getPropertyObject` or `getPropertyAsObject` for the similarity (but still important difference) with the core methods.

tlambert03 replied:
I like `getPropertyObject` ... let's go with that

tlambert03 commented:
ok, removed all the core setting stuff.  Added `core.getPropertyObject` and `core.iterProperties()` (which is an iterator i've wanted for a while... it also includes `as_object`, `device_type`, and `property_type` filters.)

ianhi commented on pymmcore_plus/core/_property.py:
```diff
@@ -0,0 +1,169 @@
+    def dict(self) -> InfoDict:
+        """Return dict of info about this Property.
+
+        Contains the following keys (See `InfoDict` type): "valid", "value", "type",
```
Is there any way for a user to access the `InfoDict` type, especially given being hidden insid ethe type_checking if?

tlambert03 replied:
Not as an import outside of a Type checking block... (but they could import inside of a type_checking block too).  what sort of usage are you picturing?

tlambert03 replied:
moved outside the type_checking clause just in case

ianhi reviewed (approved):
Awesome! This has been a real pain point for me at times and this is clearly better.

Only remaining thought is about `getDeviceProperties`. It would be nice to have the ability to get all the properties as objects for a specific device. Either as a new method (`iterDeviceProperties`?) or with an `as_object` flag.

But that shouldn't hold up this PR at all.

tlambert03 commented:
> It would be nice to have the ability to get all the properties as objects for a specific device.

added a `device_label` filter to `iterProperties()`

tlambert03 commented:
> Only remaining thought is about getDeviceProperties

actually I'd propose we remove that method.  can do in a later PR though
